Comment機能の追加

tatsukishiga opened the pull request:
１，モデルの作成
２，モデルにバリデーションの追加
３，DB側にバリデーション追加
４，アソシエーションの追加

tatsukishiga committed:
Comment機能の追加

１，モデルの作成
２，モデルにバリデーションの追加
３，DB側にバリデーション追加
４，アソシエーションの追加

tatsukishiga committed:
ルーティング、コントローラー、ビューを作成

1,ルーティングはResourcesで定義。
2,コントローラー内に必要なアクションを定義。
3,コメント編集ビューを作成。

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+      user_id: @current_user.id,
+      topic_id: params[:topic_id]
+    )
+    @comment.save
```
保存に失敗することもあるので、その処理はしておきましょう

tatsukishiga replied:
こちらrenderを使用して、失敗時の処理をしたいのですが、検証中です。
相談の詳細をSlackの方で送らさせてもらってます。

tatsukishiga replied:
完了してます

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+      topic_id: params[:topic_id]
+    )
+    @comment.save
+    redirect_to("/topics")
```
何回も言いますが、名前付きパスで！

tatsukishiga replied:
修正しました。

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  def create
+    @comment = Comment.new(
```
これも前回言いましたが

```
@current_user.comments.new(~~~)
```

とかでuser_idがセットされた状態になります

tatsukishiga replied:
修正しました

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+class CommentsController < ApplicationController
```
ログイン、未ログインは考慮してます？

tatsukishiga replied:
ログイン時のみ可能、またそのコメントの投稿者でないと編集出来ないようにしました。

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def update
+    @comment = Comment.find_by(id: params[:id])
+    @comment.update(content: params[:comment][:content])
```
ここも同じことしています

tatsukishiga replied:
重複を削除しました。

tatsukishiga committed:
指摘事項修正

１，コメント投稿失敗時のrender処理を書きましたが、まだ検証中です。
２，URL直打対策でbefore_actionを追加しました。
３，updateメソッドの重複を削除しました。
４，名前付きpathに統一しました。
５，createアクションで@current_user.comments.new(~~~)
の形に修正しました。

tatsukishiga committed:
指摘事項修正＋気づいた点修正

1,コメントの投稿に失敗した際の処理を追加。
2,redirect_backを学んだので活用出来る箇所はこちらに変更。
3,redirect_toでの飛ばし先を変更
4,ensure_correct_userメソッドが思い通りに機能していなかったため修正を加えた。

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,54 @@
+      redirect_back(fallback_location: topics_path)
+      return
+    else
+      flash[:notice] = "投稿に失敗しました。投稿が空白または、140字を超えています。"
```
投稿が空白または、140字を超えています
そのエラーとは限らなくないですかね？

tatsukishiga replied:
確かに、それ以外のエラーである可能性もあります。
エラーメッセージの表示方法として
ビューに書くエラーフルメッセージを
redirectの方でも試したのですが、
エラーが出てしまい、最終的に
こちらのコードを書きました。
今までは
仕組みとして、送られてきたフォームが
falseだった場合にエラー文が格納され
renderでそのまま表示しているという認識だったのですが、
redirectではエラーメソッドは
使用できないでしょうか？

ArakawaShintaro replied:
errorsに格納してrenderでエラーメッセージを表示させてあげるのがいいかな、と思います

tatsukishiga replied:
はい！そちらで、検証します！

tatsukishiga replied:
完了してます

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,54 @@
+    else
+      flash[:notice] = "投稿に失敗しました。投稿が空白または、140字を超えています。"
+      redirect_to(topic_path(@comment.topic_id))
+      return
```
returnいりますか？

tatsukishiga replied:
不要です。修正します！

tatsukishiga replied:
完了してます

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def edit
+    @comment = Comment.find_by(id: params[:id])
```
ensure_correct_userの部分ですが

@comment.user_id == current_user.id
で、できそうです

tatsukishiga replied:
なるほどです！修正します！

tatsukishiga replied:
完了してます

tatsukishiga committed:
指摘事項修正

1,投稿失敗時のredirectをrenderに修正。
2,createアクションでcontent値が受け取れてなかったので修正。
3,ensure_correct_userアクションの修正
4,不要なreturnの削除

ArakawaShintaro commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    else
+      flash[:notice] = "権限がありません"
+      redirect_to(topics_path)
+    end
```
```ruby
if @comment.user_id != @current_user.id
  flash[:notice] = "権限がありません"
  redirect_to(topics_path)
end
```

ArakawaShintaro replied:
private以下に追記

tatsukishiga replied:
修正しました。

tatsukishiga committed:
指摘事項修正

1,privateメソッド追加
2,if文を修正

ArakawaShintaro reviewed (approved):
OK

tatsukishiga committed:
Merge branch 'master' into comment_function
